FlexMeter.c

```diff
@@ -0,0 +1,191 @@
+/*
+htop - FlexMeter.c
+(C) 2024 Stoyan Bogdanov
+Released under the GNU GPLv2+, see the COPYING file
+in the source distribution for its full text.
+*/
+
+#include <sys/time.h>
+#include <sys/types.h>
+#include <unistd.h>
+#include <dirent.h>
+#include <time.h>
+#include <pwd.h>
+#include "FlexMeter.h"
+#include "Object.h"
+#include "config.h"
+#include "CRT.h"
+
+#define FLEX_CFG_FOLDER ".config/htop/FlexMeter"
+
+typedef struct {
+   char* name;
+   char* command;
+   char* type;
+   char* caption;
+   char* uiName;
+} _flex_meter;
+
+_flex_meter meter_list[METERS_LIST_SIZE];
+
+int meter_list_idx = 0;
+static int meters_count = 0;
+
+static const int DateMeter_attributes[] = {
+   FLEX
+};
+
+MeterClass* FlexMeter_class = NULL;
+
+int check_for_meters(void);
+
+static int parse_input(char* line)
+{
+   if (String_startsWith(line, "name=")) {
+      meter_list[meters_count].uiName = xStrdup(line + 5);
+   } else if (String_startsWith(line, "command=")) {
+      meter_list[meters_count].command = xStrdup(line + 8);
+   } else if (String_startsWith(line, "caption=")) {
+      meter_list[meters_count].caption = xStrdup(line + 8);
+   } else if (String_startsWith(line, "type=")) {
+      meter_list[meters_count].type = xStrdup(line + 6);
+   } else {
+      return -1;
+   }
+
+   return 0;
+}
+
+static int load_config(char* file)
+{
+   int ret = -1;
+   FILE* fp = fopen(file, "r");
+
+   if (fp != NULL) {
+      char* buff;
+      while ((buff = String_readLine(fp)) != NULL) {
+         ret = parse_input(buff);
+         if (ret) {
+            break;
+         }
+      }
+      free(buff);
+      buff = NULL;
+   }
+
+   fclose(fp);
+   return ret;
+}
+
+int check_for_meters(void)
+{
+   char* path;
+   struct dirent* dir;
+   struct passwd* pw = getpwuid(getuid());
+   const char* homedir = pw->pw_dir;
+   char* home = NULL;
+   int ret;
+
+   xAsprintf(&home, "%s/%s", homedir, FLEX_CFG_FOLDER);
+   struct stat fileStat;
+
+   if (stat(home, &fileStat) < 0) {
+      return -1;
+   }
+
+   uint32_t uid = getuid();
+
+   if ((fileStat.st_uid == uid) && S_ISDIR(fileStat.st_mode) && ((fileStat.st_mode & 0777) == 0700)) {
+      DIR* d = opendir(home);
+      if (d) {
+         while ((dir = readdir(d)) != NULL) {
+            if ( dir->d_name[0] == '.') {
+               /* We are ignoring all files starting with . like ".Template"
+                * and "." ".." directories
+                */
+               continue;
+            }
+
+            meter_list[meters_count].name = xStrdup(dir->d_name);
+            xAsprintf(&path, "%s/%s", home, dir->d_name);
+
+            if (access(path, R_OK | W_OK | X_OK) == 0) {
```

Comment:
Do we need write access here? RX should be fine.

Reply:
no we dont need write access , but this match what we discuss for `700`

Reply:
I think it'd be best to allow both 0500 and 0700 so the user can disable write access as that is not needed for the feature to work. The check should be for the minimum we require, with 0700 being the maximum we allow (technically even 0755 would still be fine).

Reply:
@BenBE I think even 0400 would suffice. Or maybe just remove this check altogether as it is redundant.

Reply:
(Sorry for the last comment about 0400 as I misread the code. I deleted that comment.)

By the way, do we have a better API than access(2) here? access(2) is vulnerable to TOCTOU and I wish to avoid it when possible.

Reply:
An option may be to use `open(2)` with `O_EXCL` and `fstat(2)` the `fd`, before passing it on to `fdopen(3)` in `load_config`, which would need to receive the `fd` (or a `FILE*`, if we did the `fdopen(3)` in the caller).

There is even the possibility to use `faccessat(2)` with `AT_EMPTY_FILE` to emulate `access(2)` using an already open `fd`, but this is Linux-specific (read: non-portable) and requires a recent (>=5.8) kernel to work.

Reply:
what about replacing `access(2)` with `stat(2)` like for folder ?

Reply:
@BenBE @bogdanovs 
How about doing it in the same way that we check for `htoprc`?
`open(2)` the file with the option (`O_RDONLY | O_NOFOLLOW`); after it succeeds, `fstat(2)` the file descriptor and check its execute bit `(sb.st_mode & S_IXUSR) != 0`; after both of these are done, `fdopen(3)`.

The [`stat` structure](https://man7.org/linux/man-pages/man3/stat.3type.html) should give us enough information to check not only for modes but also owner ID and group ID. If you are able, you may try implementing [the whole logic I made in a previous comment](https://github.com/htop-dev/htop/pull/1571#issuecomment-2565338735).

Reply:
> @BenBE @bogdanovs How about doing it in the same way that we check for `htoprc`? `open(2)` the file with the option (`O_RDONLY | O_NOFOLLOW`); after it succeeds, `fstat(2)` the file descriptor and check its execute bit `(sb.st_mode & S_IXUSR) != 0`; after both of these are done, `fdopen(3)`.
> 
> The [`stat` structure](https://man7.org/linux/man-pages/man3/stat.3type.html) should give us enough information to check not only for modes but also owner ID and group ID. If you are able, you may try implementing [the whole logic I made in a previous comment](https://github.com/htop-dev/htop/pull/1571#issuecomment-2565338735).

if we agree on `stat` approach I will implement it that way

Reply:
@Explorer09 What's the difference in your suggestion compared to mine? Mine was mostly focused on the actual files, but would work similar for the directory too.

Reply:
@BenBE Probably no difference, except that the `O_EXCL` flag you mentioned won't do what you want. (`O_EXCL` is for creating the file only - it forces `open(2)` to error out if a file with the same name exists).